data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/src/test/java/org/eclipse/edc/protocol/dsp/negotiation/api/controller/DspNegotiationControllerTest.java

```diff
@@ -0,0 +1,552 @@
+        var token = token();
+        var message = contractRequestMessage();
+        var process = contractNegotiation();
+        var request = Json.createObjectBuilder().add("@type", DSPACE_NEGOTIATION_CONTRACT_REQUEST_MESSAGE).build();
+
+        when(identityService.verifyJwtToken(any(TokenRepresentation.class), eq(callbackAddress))).thenReturn(Result.success(token));
+        when(registry.transform(any(JsonObject.class), eq(ContractRequestMessage.class))).thenReturn(Result.success(message));
+        when(protocolService.notifyRequested(message, token)).thenReturn(ServiceResult.success(process));
+        when(registry.transform(any(ContractNegotiation.class), eq(JsonObject.class))).thenReturn(Result.failure("error"));
+
+        baseRequest()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(request)
+                .post(BASE_PATH + INITIAL_CONTRACT_REQUEST)
+                .then()
+                .statusCode(500);
+    }
+
+    @Test
+    void initiateNegotiation_shouldReturnInternalServerError_whenConvertingResultFails() {
+        var token = token();
+        var message = contractRequestMessage();
+        var process = contractNegotiation();
+        var json = Json.createObjectBuilder().build();
+        var request = Json.createObjectBuilder().add("@type", DSPACE_NEGOTIATION_CONTRACT_REQUEST_MESSAGE).build();
+
+        when(identityService.verifyJwtToken(any(TokenRepresentation.class), eq(callbackAddress))).thenReturn(Result.success(token));
+        when(registry.transform(any(JsonObject.class), eq(ContractRequestMessage.class))).thenReturn(Result.success(message));
+        when(protocolService.notifyRequested(message, token)).thenReturn(ServiceResult.success(process));
+        when(registry.transform(any(ContractNegotiation.class), eq(JsonObject.class))).thenReturn(Result.success(json));
+        when(mapper.convertValue(any(JsonObject.class), eq(Map.class))).thenThrow(IllegalArgumentException.class);
+
+        baseRequest()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(request)
+                .post(BASE_PATH + INITIAL_CONTRACT_REQUEST)
+                .then()
+                .statusCode(500);
+    }
+
+    @Test
+    void providerOffer_shouldReturnNotImplemented_whenOperationNotSupported() {
+        var token = token();
+
+        when(identityService.verifyJwtToken(any(TokenRepresentation.class), eq(callbackAddress))).thenReturn(Result.success(token));
+
+        baseRequest()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(request)
+                .post(BASE_PATH + "testId" + CONTRACT_OFFER)
+                .then()
+                .statusCode(501);
+    }
+
+    /**
+     * Verifies that an endpoint returns 401 if the identity service cannot verify the identity token.
+     *
+     * @param path the request path to the endpoint
+     */
+    @ParameterizedTest
+    @MethodSource("controllerMethodArguments")
```

Comment:
nit: `@ArgumentsSource` would work better because it will be bound by a class type instead of a string.